### PR TITLE
Double-clicking in LineEdit

### LINKED ISSUE #1093: Implement double click select for LineEdit

### PR DESCRIPTION
**Pull request by ficcialfaint**
Closes #1093, #4857
Only for LineEdit (no TextEdit support yet™)

---

Two clicks count as a double-click if
* The delay between them is less than or equal to **250 ms**
* The distance between the clicks is less than **10 pixels**

Double-clicking a word selects it. Double-clicking an empty place selects two words along its edges

### CODE REVIEW

**Comment by PJB3005** [comment]
God fucking bless

**Comment by ike709** [comment]
please god merge this

File: Robust.Client/UserInterface/Controls/LineEdit.cs [hunk 1]
```diff
@@ -46,6 +47,12 @@ public class LineEdit : Control
         private bool _mouseSelectingText;
         private float _lastMousePosition;
 
+        private TimeSpan? _lastClickTime;
+        private Vector2? _lastClickPosition;
+
+        private const int DoubleClickDelay = 250;
+        private const int DoubleClickRange = 10;
```
Comment (ShadowCommander) [inline Robust.Client/UserInterface/Controls/LineEdit.cs]: Are you able to make these client CVars?

File: Robust.Client/UserInterface/Controls/LineEdit.cs [hunk 2]
```diff
@@ -685,8 +692,27 @@ async void DoPaste()
                     args.Handle();
                 }
             }
+
+            // Double-clicking. Clicks delay should be <= 250ms and the distance < 10 pixels.
+            else if (args.Function == EngineKeyFunctions.UIClick && _lastClickPosition != null && _lastClickTime != null
+                     && _timing.RealTime - _lastClickTime <= TimeSpan.FromMilliseconds(DoubleClickDelay)
+                     && (_lastClickPosition.Value - args.PointerLocation.Position).Length() < DoubleClickRange)
```
Comment (ShadowCommander) [inline Robust.Client/UserInterface/Controls/LineEdit.cs]:
```suggestion
                     && (_lastClickPosition.Value - args.PointerLocation.Position).IsShorterThan(DoubleClickRange))
```
We now have methods for comparing with vectors.

**Review by ShadowCommander — CHANGES_REQUESTED** [review]

**Pull request merged**